REGRESSION (261984@main): Copied tables from some web pages don't properly paste into Numbers

whsieh opened the pull request:
#### 664af9c1e8d722f89b18f8783a1737d799a2fafc
<pre>
REGRESSION (261984@main): Copied tables from some web pages don&apos;t properly paste into Numbers
<a href="https://bugs.webkit.org/show_bug.cgi?id=259928">https://bugs.webkit.org/show_bug.cgi?id=259928</a>
rdar://112030767

Reviewed by Megan Gardner.

When copying a table in Quip and pasting into the Numbers app, tables are programmatically written
to the pasteboard as HTML; upon paste, UIFoundation uses `nsattributedstringagent` to convert this
pasted markup into an `NSAttributedString`, by loading the markup in a `WKWebView` and asking it to
`-_getContentsAsAttributedStringWithCompletionHandler:`. This exercises the `HTMLConverter` codepath
in WebCore, which implements the correct logic for representing tables and lists in the markup as
`-textBlocks` and `-textLists`, respectively, on `NSParagraphStyle`; for tables, we list multiple
`NSTextTableBlock`s that all point to the same `NSTextTable` object. For instance, two adjacent
cells in the same table would be represented as:

```
&quot;NSParagraphStyle&quot; =&gt; {
    textBlocks =&gt; [ NSTextTableBlock 0x6000006e7100, table=&lt;NSTextTable 0x6000003f27d0&gt;, {0, 0} ],
    …
}
&quot;NSParagraphStyle&quot; =&gt; {
    textBlocks =&gt; [ NSTextTableBlock 0x6000006e6d80, table=&lt;NSTextTable 0x6000003f27d0&gt;, {0, 1} ],
    …
}
```

Similarly, lists are represented by having multiple paragraph styles with the same `NSTextList`s.
Two adjacent items in the same list, for instance, will have paragraph styles whose `-textLists`
array contains the exact same list (i.e. equal pointers):

```
&quot;NSParagraphStyle&quot; =&gt; {
    textLists =&gt; [ NSTextList 0x600002676400 format &lt;{disc}&gt; ],
    …
}
&quot;NSParagraphStyle&quot; =&gt; {
    textLists =&gt; [ NSTextList 0x600002676400 format &lt;{disc}&gt; ],
    …
}
```

Prior to the changes in 261984@main, we used a single `NSKeyedArchiver`/`NSKeyedUnarchiver` to
encode/decode an `NSAttributedString` when sending this data from the web process to the UI process
for writing to the pasteboard. This meant that Foundation&apos;s internal object caching mechanism in the
keyed archiver would ensure that multiple `NSParagraphStyle` instances that all reference the same
`NSTextList` or `NSTextTable` upon encoding would still reference the same list or table upon
decoding, thereby preserving the table/list structure when writing to the pasteboard.

However, after 261984@main, we now use a different keyed (un)archiver when encoding/decoding each
individual attribute; as a result, different `NSParagraphStyle` instances that reference the same
`NSTextList` or `NSTextTable` no longer correspond to the same object in the archiver&apos;s backing map.
This means that in the two above examples, we&apos;d end up with two 2x1 tables (each with only 1 cell
populated), and two single-item lists, respectively.

To fix this, we add some logic to plumb unique identifiers corresponding to tables or lists for each
`NSTextBlock` or `NSTextList` in `-[NSParagraphStyle textBlocks]` and `-[NSParagraph textLists]`,
which are propagated along with the rest of the attributes upon encoding. When decoding, we use
these unique identifiers to ensure that all paragraph styles that previously referenced the same
table or list will continue to do so after decoding.

Tests:  WKWebView.AttributedStringFromTable
        WKWebView.AttributedStringFromList

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:

Move various UIKit SPI and IPI declarations out of the implementation file in `HTMLConverter.mm`
and into `UIKitSPI.h` instead, so that these declarations can be used in both `AttributedString.mm`
and `HTMLConverter.mm`.

* Source/WebCore/editing/cocoa/AttributedString.h:

Instead of encoding a single `RetainPtr&lt;NSParagraphStyle&gt;`, send a struct containing an
`RetainPtr&lt;NSParagraphStyle&gt;`, a list of `TextTableID` representing each text block, and a list of
`TextListID`. Note that `tableIDs` is a list of optional identifiers, since a `textBlock` may not
necessarily correspond to a table cell, in which case we use represent it via `nullopt`. `textLists`
doesn&apos;t have this same constraint since each entry corresponds directly to an `NSTextList`.

* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::reconstructStyle):

Add a helper method to take `HashMap`s of tables and lists by ID, and (only if necessary) create a
copy of the decoded `NSParagraphStyle` whose referenced tables and lists match those that were
previously referenced when decoding earlier parts of the attributed string.

(WebCore::toNSObject):
(WebCore::toNSDictionary):

Maintain maps of tables and lists by unique ID over the whole lifetime of decoding the string.

(WebCore::AttributedString::documentAttributesAsNSDictionary const):
(WebCore::AttributedString::nsAttributedString const):
(WebCore::extractListIDs):
(WebCore::extractTableIDs):
(WebCore::extractValue):
(WebCore::extractDictionary):
(WebCore::AttributedString::fromNSAttributedStringAndDocumentAttributes):

Implement the opposite half of the above logic, by collecting identical `NSTextList`s and
`NSTextTable`s that are referenced by `NSParagraphStyle`s during encoding. This allows us to build a
list of unique IDs that correspond to each table or list, which we send over IPC and use during
decoding to preserve references to the same objects when deserializing paragraph styles.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
(WebCore::extensionsForMIMETypeMap):

Drive-by fix: suppress a new deprecation warning.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(-[WKWebView _contentsAsAttributedString]):

Add a couple of API tests to exercise attributed string conversion for tables and lists.
Importantly, these tests verify that table/list structure is preserved in the decoded string.

Canonical link: <a href="https://commits.webkit.org/266700@main">https://commits.webkit.org/266700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aca85bc729a6fff40c051726c82f2285619bfd40

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14477 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14787 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16223 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13692 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14612 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14864 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14658 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15186 "2 new passes 9 flakes 1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16962 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12471 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13057 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20077 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13548 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13222 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16458 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13776 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11608 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13063 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3511 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17400 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -40,10 +41,10 @@
 #define PlatformNSShadow                NSShadow.class
 #define PlatformNSTextAttachment        NSTextAttachment.class
-#define PlatformNSTextList              NSTextList
+#define PlatformNSTextList              NSTextList.class
```
Just curious, why is this .class needed now?

whsieh replied:
Oh, good catch! We only need this change for the `PlatformNSTextTableBlock` define, not text lists.

whsieh committed:
REGRESSION (261984@main): Copied tables from some web pages don't properly paste into Numbers
https://bugs.webkit.org/show_bug.cgi?id=259928
rdar://112030767

Reviewed by Megan Gardner.

When copying a table in Quip and pasting into the Numbers app, tables are programmatically written
to the pasteboard as HTML; upon paste, UIFoundation uses `nsattributedstringagent` to convert this
pasted markup into an `NSAttributedString`, by loading the markup in a `WKWebView` and asking it to
`-_getContentsAsAttributedStringWithCompletionHandler:`. This exercises the `HTMLConverter` codepath
in WebCore, which implements the correct logic for representing tables and lists in the markup as
`-textBlocks` and `-textLists`, respectively, on `NSParagraphStyle`; for tables, we list multiple
`NSTextTableBlock`s that all point to the same `NSTextTable` object. For instance, two adjacent
cells in the same table would be represented as:

```
"NSParagraphStyle" => {
    textBlocks => [ NSTextTableBlock 0x6000006e7100, table=<NSTextTable 0x6000003f27d0>, {0, 0} ],
    …
}
"NSParagraphStyle" => {
    textBlocks => [ NSTextTableBlock 0x6000006e6d80, table=<NSTextTable 0x6000003f27d0>, {0, 1} ],
    …
}
```

Similarly, lists are represented by having multiple paragraph styles with the same `NSTextList`s.
Two adjacent items in the same list, for instance, will have paragraph styles whose `-textLists`
array contains the exact same list (i.e. equal pointers):

```
"NSParagraphStyle" => {
    textLists => [ NSTextList 0x600002676400 format <{disc}> ],
    …
}
"NSParagraphStyle" => {
    textLists => [ NSTextList 0x600002676400 format <{disc}> ],
    …
}
```

Prior to the changes in 261984@main, we used a single `NSKeyedArchiver`/`NSKeyedUnarchiver` to
encode/decode an `NSAttributedString` when sending this data from the web process to the UI process
for writing to the pasteboard. This meant that Foundation's internal object caching mechanism in the
keyed archiver would ensure that multiple `NSParagraphStyle` instances that all reference the same
`NSTextList` or `NSTextTable` upon encoding would still reference the same list or table upon
decoding, thereby preserving the table/list structure when writing to the pasteboard.

However, after 261984@main, we now use a different keyed (un)archiver when encoding/decoding each
individual attribute; as a result, different `NSParagraphStyle` instances that reference the same
`NSTextList` or `NSTextTable` no longer correspond to the same object in the archiver's backing map.
This means that in the two above examples, we'd end up with two 2x1 tables (each with only 1 cell
populated), and two single-item lists, respectively.

To fix this, we add some logic to plumb unique identifiers corresponding to tables or lists for each
`NSTextBlock` or `NSTextList` in `-[NSParagraphStyle textBlocks]` and `-[NSParagraph textLists]`,
which are propagated along with the rest of the attributes upon encoding. When decoding, we use
these unique identifiers to ensure that all paragraph styles that previously referenced the same
table or list will continue to do so after decoding.

Tests:  WKWebView.AttributedStringFromTable
        WKWebView.AttributedStringFromList

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:

Move various UIKit SPI and IPI declarations out of the implementation file in `HTMLConverter.mm`
and into `UIKitSPI.h` instead, so that these declarations can be used in both `AttributedString.mm`
and `HTMLConverter.mm`.

* Source/WebCore/editing/cocoa/AttributedString.h:

Instead of encoding a single `RetainPtr<NSParagraphStyle>`, send a struct containing an
`RetainPtr<NSParagraphStyle>`, a list of `TextTableID` representing each text block, and a list of
`TextListID`. Note that `tableIDs` is a list of optional identifiers, since a `textBlock` may not
necessarily correspond to a table cell, in which case we use represent it via `nullopt`. `textLists`
doesn't have this same constraint since each entry corresponds directly to an `NSTextList`.

* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::reconstructStyle):

Add a helper method to take `HashMap`s of tables and lists by ID, and (only if necessary) create a
copy of the decoded `NSParagraphStyle` whose referenced tables and lists match those that were
previously referenced when decoding earlier parts of the attributed string.

(WebCore::toNSObject):
(WebCore::toNSDictionary):

Maintain maps of tables and lists by unique ID over the whole lifetime of decoding the string.

(WebCore::AttributedString::documentAttributesAsNSDictionary const):
(WebCore::AttributedString::nsAttributedString const):
(WebCore::extractListIDs):
(WebCore::extractTableIDs):
(WebCore::extractValue):
(WebCore::extractDictionary):
(WebCore::AttributedString::fromNSAttributedStringAndDocumentAttributes):

Implement the opposite half of the above logic, by collecting identical `NSTextList`s and
`NSTextTable`s that are referenced by `NSParagraphStyle`s during encoding. This allows us to build a
list of unique IDs that correspond to each table or list, which we send over IPC and use during
decoding to preserve references to the same objects when deserializing paragraph styles.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
(WebCore::extensionsForMIMETypeMap):

Drive-by fix: suppress a new deprecation warning.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(-[WKWebView _contentsAsAttributedString]):

Add a couple of API tests to exercise attributed string conversion for tables and lists.
Importantly, these tests verify that table/list structure is preserved in the decoded string.

Canonical link: https://commits.webkit.org/266700@main
